File: reasoner/clingo_engine.py
"""Clingo ASP engine wrapper.

Provides two modes:
- Datalog (deterministic): single answer set for transitive inference
- ASP (non-deterministic): multiple answer sets for conflict resolution
"""

import logging
from pathlib import Path
from typing import TYPE_CHECKING

from clingo import Control

if TYPE_CHECKING:
    from extractors.triple_schema import ExtractedTriple

logger = logging.getLogger(__name__)

RULES_DIR = Path(__file__).parent
RULES_ASP = str(RULES_DIR / "rules.asp")
CONFLICT_ASP = str(RULES_DIR / "asp_conflict.asp")


def triples_to_asp_facts(triples: list["ExtractedTriple"]) -> str:
    """Convert extracted triples into Clingo ASP fact strings.

    Args:
        triples: List of ExtractedTriple objects

    Returns:
        Multi-line string of ASP facts
    """
    lines: list[str] = []
    for t in triples:
        s = t.subject.replace('"', "'")
        o = t.obj.replace('"', "'")
        src = t.source_paper.replace('"', "'")
        # e.g.: usa("bert", "transformer").
        lines.append(f'{t.predicate}("{s}", "{o}").  % src={src} conf={t.confidence:.2f}')
    return "\n".join(lines)


class ClingoEngine:
    """Wrapper around Clingo for Datalog and ASP reasoning."""

    def run_datalog(self, facts: str) -> list[str]:
        """Run deterministic Datalog inference.

        Returns a single answer set's atoms as strings.
        Raises ValueError if the program is unsatisfiable.
        """
        ctl = Control()
        ctl.add("base", [], facts)
        ctl.load(RULES_ASP)
        ctl.ground([("base", [])])

        inferred: list[str] = []

        def collect_model(model) -> None:
            inferred.extend(str(sym) for sym in model.symbols(shown=True))

        result = ctl.solve(on_model=collect_model)

        if result.unsatisfiable:
            raise ValueError("Datalog program is unsatisfiable — check for contradictory facts")

        logger.info("clingo datalog: %d atoms inferred", len(inferred))
        ctl.cleanup()
        return inferred

    def run_asp(self, facts: str, max_models: int = 10) -> list[list[str]]:
        """Run non-deterministic ASP conflict resolution.

        Returns multiple answer sets — each represents a valid interpretation.
        """
        ctl = Control(["0"])
        ctl.configuration.solve.models = max_models
        ctl.add("base", [], facts)
        ctl.load(RULES_ASP)
        ctl.load(CONFLICT_ASP)
        ctl.ground([("base", [])])

        answer_sets: list[list[str]] = []

        with ctl.solve(yield_=True) as handle:
            for model in handle:
                answer_sets.append([str(sym) for sym in model.symbols(shown=True)])

        logger.info("clingo asp: %d answer sets found", len(answer_sets))
        ctl.cleanup()
        return answer_sets

    def has_conflicts(self, inferred: list[str]) -> bool:
        """Check if any conflict_potencial atoms exist in the inferred facts."""
        return any("conflito_potencial" in atom for atom in inferred)

    def extract_conflicts(self, inferred: list[str]) -> list[str]:
        """Return only the conflict atoms from inferred facts."""
        return [atom for atom in inferred if "conflito_potencial" in atom]

File: src/reasoner/clingoEngine.ts
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";
import type { Triple } from "../graph/state.js";
import { getLogger } from "../observability.js";

const log = getLogger("reasoner:clingo");

const __dirname = dirname(fileURLToPath(import.meta.url));

// clingo-wasm module reference — initialized once
let clingoReady = false;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ClingoModule = { run: (program: string, models: number) => Promise<ClingoResult> };

interface ClingoWitness {
  Value: string[];
  Time?: number;
}

interface ClingoCall {
  Witnesses: ClingoWitness[];
}

interface ClingoResult {
  Result: "SATISFIABLE" | "UNSATISFIABLE" | "UNKNOWN" | "ERROR";
  Call?: ClingoCall[];
  Error?: string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let clingoMod: any;

async function getClingo(): Promise<ClingoModule> {
  if (!clingoReady) {
    // clingo-wasm exports: { init, run, Runner }
    const mod = await import("clingo-wasm");
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    clingoMod = mod as any;
    await clingoMod.init();
    clingoReady = true;
  }
  return clingoMod as ClingoModule;
}

/** Convert triples to ASP facts */
function triplesToFacts(triples: Triple[]): string {
  return triples
    .map((t) => {
      const subj = t.subject.toLowerCase().replace(/\s+/g, "_").replace(/[^a-z0-9_]/g, "");
      const obj = t.object.toLowerCase().replace(/\s+/g, "_").replace(/[^a-z0-9_]/g, "");
      if (!subj || !obj) return null;
      return `${t.predicate}(${subj}, ${obj}).`;
    })
    .filter(Boolean)
    .join("\n");
}

function loadRules(filename: string): string {
  try {
    return readFileSync(join(__dirname, filename), "utf8");
  } catch {
    log.warn({ filename }, "Rules file not found, using empty rules");
    return "";
  }
}

function extractAtoms(result: ClingoResult): string[] {
  return result.Call?.[0]?.Witnesses?.[0]?.Value ?? [];
}

export class ClingoEngine {
  /** Deterministic Datalog inference — single stable model */
  async runDatalog(
    triples: Triple[],
  ): Promise<{ inferredFacts: string[]; conflicts: string[] }> {
    const clingo = await getClingo();
    const facts = triplesToFacts(triples);
    const rules = loadRules("rules.lp");
    const program = `${facts}\n${rules}`;

    log.debug({ facts: triples.length }, "Running Datalog inference");

    const result = await clingo.run(program, 1);

    if (result.Result === "UNSATISFIABLE" || result.Result === "ERROR") {
      log.warn({ result: result.Result, error: result.Error }, "Datalog issue");
      return { inferredFacts: [], conflicts: [] };
    }

    const atoms = extractAtoms(result);
    const inferredFacts = atoms.filter((a) => a.startsWith("inferred("));
    const conflicts = atoms.filter((a) => a.startsWith("conflict("));

    return { inferredFacts, conflicts };
  }

  /** Non-deterministic ASP — enumerate answer sets for conflict resolution */
  async runAsp(triples: Triple[], maxModels = 10): Promise<string[][]> {
    const clingo = await getClingo();
    const facts = triplesToFacts(triples);
    const rules = loadRules("rules.lp");
    const conflictRules = loadRules("aspConflict.lp");
    const program = `${facts}\n${rules}\n${conflictRules}`;

    log.debug({ maxModels }, "Running ASP conflict resolution");

    const result = await clingo.run(program, maxModels);

    if (result.Result === "ERROR") {
      log.warn({ error: result.Error }, "ASP error");
      return [];
    }

    return (result.Call?.[0]?.Witnesses ?? []).map((w: ClingoWitness) => w.Value ?? []);
  }
}
